shared/data/web_accessible_resources/youtube-iframe-api.js

```diff
@@ -0,0 +1,288 @@
+
+    // Mappings between the "real" video elements and their placeholder
+    // elements.
+    const videoElementsByID = new Map()
+    const videoElementByPlaceholderElement = new WeakMap()
+    const placeholderElementByVideoElement = new WeakMap()
+
+    /**
+     * Mock of the `YT.Player` constructor.
+     */
+    function Player (target, config = { }, ...rest) {
+        if (youTubeIframeAPILoaded) {
+            return new RealYTPlayer(target, config, ...rest)
+        }
+
+        const { height, width, videoId, playerVars = { }, events } = config
+
+        if (!(target instanceof Element)) {
+            target = document.getElementById(target)
+        }
+
+        // Normalise target to always be the video element instead of the
+        // placeholder element (if either even exists at this point).
+        if (videoElementByPlaceholderElement.has(target)) {
+            target = videoElementByPlaceholderElement.get(target)
+        }
+
+        if (!target) {
+            throw new Error('Target not found')
+        }
+
+        // Set up the video element if the target isn't an existing video.
+        if (!placeholderElementByVideoElement.has(target)) {
+            const url = new URL(window.YTConfig.host)
+            url.pathname = '/embed/'
+
+            // For videos (not playlists) append the video ID to the path.
+            // See https://developers.google.com/youtube/player_parameters
+            if (!playerVars.list) {
+                url.pathname += encodeURIComponent(videoId)
+            }
+
+            for (const [key, value] of Object.entries(playerVars)) {
+                url.searchParams.set(key, value)
+            }
+
+            // Ensure that JavaScript control of the video is enabled. This is
+            // necessary for the onReady event to fire for the video.
+            url.searchParams.set('enablejsapi', '1')
+
+            const videoIframe = document.createElement('iframe')
+            videoIframe.height = parseInt(height, 10) || defaultHeight
+            videoIframe.width = parseInt(width, 10) || defaultWidth
+            videoIframe.src = url.href
+
+            if (target.id) {
+                videoIframe.id = target.id
+            }
+
+            target.replaceWith(videoIframe)
+            target = videoIframe
+
+            target.dispatchEvent(new CustomEvent('ddg-ctp-replace-element'))
+        }
+
+        if (events) {
+            if (events.onReady) {
+                onReadyListenerByVideoElement.set(target, events.onReady)
+            }
+            if (events.onStateChange) {
+                onStateChangeListenerByVideoElement.set(target, events.onStateChange)
+            }
+        }
+
+        mockPlayerByVideoElement.set(target, this)
+        this.playerInfo = {}
+        return this
+    }
+
+    // Stub out the YouTube Iframe API.
+    window.YTConfig = {
+        host: 'https://www.youtube.com'
+    }
+    window.YT = {
+        loading: 1,
```

Comment:
I'm wondering if we should have schema unit tests on the real object and compare it to these mocks? That way if the API changes the build breaks etc.

Reply:
Yea, good idea.

Reply:
This review is large enough, so I've split that out into a separate Asana ticket. Hopefully I'll get a chance to do that during the quick win days.